### PR TITLE
Add LocalElementMap

### PR DESCRIPTION
**Pull request by simonbyrne**
This refactors the mesh a bit to support a `LocalElementMap` option.

### CODE REVIEW

**Comment by simonbyrne** [comment]
It might be good to merge this in for now. I can do the interpolation and higher level pieces in a new PR.

File: src/Geometry/coordinates.jl [hunk 1]
```diff
@@ -240,6 +240,64 @@ function bilinear_interpolate(
...
+    Δ = c1 * c1 - 4 * c2 * c0
+    if c1 >= 0
+        ξ2a = (-c1 - sqrt(Δ)) / (2 * c2)
+        ξ2b = 2 * c0 / (-c1 - sqrt(Δ))
+    else
+        ξ2a = 2 * c0 / (-c1 + sqrt(Δ))
+        ξ2b = (-c1 + sqrt(Δ)) / (2 * c2)
+    end
+    # 7) find which solution is valid
+    if abs(ξ2a) < 1.00001
```
Comment (jakebolewski) [inline src/Geometry/coordinates.jl:290]: how was this tolerance chosen?

File: src/Meshes/cubedsphere.jl [hunk 2]
```diff
@@ -137,27 +163,162 @@ function panel_to_coordinates(panel::Integer, coord::Geometry.Cartesian123Point)
...
 """
-    panel, coord1 = coordinates_to_panel(coord::Geometry.Cartesian123Point)
+    panel = cubedspherepanel(coord::Geometry.Cartesian123Point)
 
-Given a point `coord`, return its panel number (`panel`), and its coordinates in panel 1
-(`coord1`).
+Given a point `coord`, return its panel number (an integer between 1 and 6).
 """
-function coordinates_to_panel(coord::Geometry.Cartesian123Point)
+function containing_panel(coord::Geometry.Cartesian123Point)
```
Comment (jakebolewski) [inline src/Meshes/cubedsphere.jl:170]: :+1:

**Review by jakebolewski — APPROVED** [review]

File: src/Spaces/spectralelement.jl [hunk 3]
```diff
@@ -155,112 +155,45 @@ function SpectralElementSpace2D(topology, quadrature_style)
     quad_points, quad_weights =
         Quadratures.quadrature_points(FT, quadrature_style)
 
-    for elem in 1:nelements
-        local_geometry_slab = slab(local_geometry, elem)
+    for e in 1:nelements
```
Comment (jakebolewski) [inline src/Spaces/spectralelement.jl:158]: :+1:

File: src/Geometry/coordinates.jl (hunk 1 above)
Comment (simonbyrne) [reply src/Geometry/coordinates.jl:290]: very rigorously 😄

Comment (simonbyrne) [reply src/Geometry/coordinates.jl:290]: Actually, we can probably just choose the smallest one.

**Pull request merged**